lumicks/pylake/fitting/profile_likelihood.py

```diff
@@ -0,0 +1,333 @@
+    step_function: callable
+        function which performs 1D line scans
+    termination_level: float
+        chi squared value at which the optimization terminates
+    """
+    lower_bounds: np.array
+    upper_bounds: np.array
+    fitted: np.array
+    step_function: callable
+    termination_level: float
+
+
+class ProfileInfo(NamedTuple):
+    minimum_chi2: float
+    profiled_parameter_index: int
+    delta_chi2: float
+    confidence_level: float
+    parameter_names: list
+
+
+def do_step(chi2_function, step_direction_function, chi2_last, parameter_vector, current_step_size, step_sign,
+            step_config):
+    """
+    Parameters
+    ----------
+    chi2_function: callable
+        function which returns the current chi2 value
+    step_direction_function: callable
+        function which returns normalized direction in parameter space in which steps are taken
+    chi2_last: float
+        previous chi squared value
+    parameter_vector: array_like
+        current parameter vector
+    current_step_size: float
+        current step size
+    step_sign: float
+        direction in which to step (-1.0 or 1.0)
+    step_config: StepConfig
+        configuration for the stepping algorithm
+    """
+    # Determine an appropriate step size based on chi2 increase
+    adjust_trial = True
+    just_shrunk = False
+    current_step_size = np.min([current_step_size, step_config.max_abs_step])
+
+    step_direction = step_direction_function()
+    while adjust_trial:
+        p_trial, clamped = clamp_step(parameter_vector, step_sign * current_step_size * step_direction,
+                                      step_config.lower_bounds, step_config.upper_bounds)
+
+        just_shrunk = just_shrunk or clamped
+        chi2_trial = chi2_function(p_trial)
+
+        chi2_change = chi2_trial - chi2_last
+        if chi2_change < step_config.min_chi2_step_size:
+            # Do not increase the step-size if we just shrunk. We already know it's going to be bad and we'd
+            # just be looping forever.
+            if not just_shrunk:
+                adjust_trial = True
+                current_step_size = current_step_size * step_config.step_factor
+                if current_step_size > step_config.max_abs_step:
+                    current_step_size = step_config.max_abs_step
+                    adjust_trial = False
+            else:
+                adjust_trial = False
+        elif chi2_change > step_config.max_chi2_step_size:
+            adjust_trial = True
+            just_shrunk = True
+            current_step_size = current_step_size / step_config.step_factor
+            if current_step_size < step_config.min_abs_step:
+                warn("Warning: Step size set to minimum step size.", RuntimeWarning)
+                current_step_size = step_config.min_abs_step
+                adjust_trial = False
+        else:
+            adjust_trial = False
+            just_shrunk = False
```

Comment:
Yeah, sorry about this, this loop got updated in the latest version and I should've squashed that commit down with the one that introduced it. I think the loop you suggest does something different as that one doesn't verify that the step size isn't too big if we hit the max step (it should verify that the `chi2` doesn't go up too much, and otherwise shrink again).

I think if you cover `no change` and `grow after shrink`, you can't get an infinite loop. I'm not sure how I feel about adding a maximum iteration count for this, since it's a parameter that shouldn't have to exist.